src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -292,8 +319,131 @@ internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort a
                     _ => hex
                 };
 
-                return new MappedField(name, this, type, key.Data);
-            });
+                return new MappedField(name, @this, type, data);
+            },
+            this);
+        }
+
+        internal DataSectionStringType? GetOrCreateDataSectionStringLiteralType(
+            string text,
+            ImmutableArray<byte> data,
+            Compilation compilation,
+            SyntaxNode syntaxNode,
+            DiagnosticBag diagnostics)
+        {
+            return _dataSectionStringLiteralTypes.GetOrAdd(text, static (key, arg) =>
+            {
+                var (@this, data, compilation, syntaxNode, diagnostics) = arg;
+
+                string name = "<S>" + DataToHexViaXxHash128(data);
+
+                MappedField dataField = @this.CreateDataField(data, alignment: 1);
+
+                Cci.IMethodDefinition bytesToStringHelper = @this.GetOrSynthesizeBytesToStringHelper(compilation, syntaxNode, diagnostics);
+
+                return new DataSectionStringType(
+                    module: (ITokenDeferral)@this._moduleBuilder,
+                    name: name,
+                    systemObject: @this._systemObject,
+                    systemString: @this._systemStringType,
+                    containingType: @this,
+                    dataField: dataField,
+                    bytesToStringHelper: bytesToStringHelper,
+                    diagnostics: diagnostics);
+            },
+            (this, data, compilation, syntaxNode, diagnostics));
+        }
+
+        /// <summary>
+        /// Gets the <see cref="BytesToStringHelper"/> or creates it if it does not exist yet.
+        /// </summary>
+        private Cci.IMethodDefinition GetOrSynthesizeBytesToStringHelper(Compilation compilation, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
+        {
+            var method = GetMethod(SynthesizedBytesToStringFunctionName);
+
+            if (method is null)
+            {
+                TryAddSynthesizedMethod(synthesizeBytesToStringHelper(this, compilation, syntaxNode, diagnostics));
+
+                method = GetMethod(SynthesizedBytesToStringFunctionName);
+            }
+
+            Debug.Assert(method is not null);
+            return method;
+
+            static Cci.IMethodDefinition synthesizeBytesToStringHelper(
+                PrivateImplementationDetails @this,
+                Compilation compilation,
+                SyntaxNode syntaxNode,
+                DiagnosticBag diagnostics)
+            {
+                if (!tryGetWellKnownTypeMember(compilation, syntaxNode, diagnostics, WellKnownMember.System_Text_Encoding__get_UTF8, out var encodingUtf8) |
+                    !tryGetWellKnownTypeMember(compilation, syntaxNode, diagnostics, WellKnownMember.System_Text_Encoding__GetString, out var encodingGetString))
+                {
+                    return new DefaultMethodDef(
```

Comment:
>return new DefaultMethodDef( [](http://example.com/codeflow?start=20&length=28)

I am not sure what is going on here. Are we creating a "bogus" method definition in order to recover from an error? I think it would be preferred to not do that.

Reply:
We can make it a responsibility of the only consumer of this functionality, the ```TryEmitStringLiteralAsUtf8Encoded```, to check the presence/goodness of required APIs and to report appropriate diagnostics. In this file, we could assume that relevant APIs won't be called unless the helpers are good. I think if we do that, the logic in this file can be simplified, changes to ```CommonMessageProvider``` can be reverted, etc.